Create final demand imputation output notebook

zschira opened the pull request:
# Overview

Updates 930 imputation notebook to contain more narrative and flow, and consolidates all useful visualizations in one notebook.

zaneselvans commented:
<p>I think it would be good to have at least 1 example from the FERC-714 data.</p><p> </p><p>One "practical" example that's easy to do is highlighting the drop in electricity demand during the initial pandemic lockdowns.</p>

---

 <a href='https://app.reviewnb.com/catalyst-cooperative/pudl-examples/pull/10/discussion/#comment-3059004845'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

zaneselvans reviewed (changes requested):
Okay well, I can't figure out how to submit my comments / review on n ReviewNB, so I will cut-and-paste here.  Main issues:

- This python environment won't exist on Kaggle.
- One of the example plots seems not to be the one you're referring to.
- The PG&E imputation discrepancy looks like a genuine failure on our part.
- Agree that another "practical" example that's not related to validating the imputation would be useful... a silly example I've used in the past is trying to see the drop in demand resulting from march/april 2020 pandemic lockdowns. There's probably something better though.

## Python environment
I don't think mamba (or conda?) will work on Kaggle, but you can do pip installs, and you can specify additional dependencies through the Kaggle interface.  However, We can't rely on installing the pudl package on Kaggle or in the user's local environment -- it has like 600 indirect dependencies and I think we're likely to stop distributing it as a package altogether, so anything in the notebook needs to rely only on packages that are part of a "standard" data science environment (i.e. whatever Kaggle has installed) -- we can also specify some additional packages to be added to the notebook's environment in association with the notebook on Kaggle, but that information will not be available to a user who checks out the notebook from GitHub. We could have a separate environment spec in the github repo, but would need to figure out how to keep it synchronized with Kaggle for both of them to work. 

## Accessing remote data
Is there a reason not to rely on the shorter and more readable S3 URLs rather than the https version? I thought there were also additional features of the S3 protocol that aren't available through https.  In the python cell it's pointing at S3, but in the Markdown it's pointing at https.

## Referencing a different plot?
In the 2nd example of heavily imputed regions: I don't 'think the imputed values are NULL. It looks more like `local_outlier_high` and `local_outlier_low`. Is the selection of these non-deterministic? I see that you've set a random state, but it seems like maybe this one changed.

## Our PG&E imputation is... wrong?
Looking at the final plot of 2019 PG&E demand, which was an outlier on the old vs. new imputation results, this is interesting and also... bad? Looking at the diurnal variation in demand that's present across the rest of the PG&E data for this year, our imputation seems implausible -- why would there be so little variability across those days of demand? What is our method keying of of here that's yielding an implausible result? Can we identify any other substantial chunks of imputation where we're consistently far from the other imputation method? This seems like a failure of our imputation method that requires explanation or correction.

zschira commented:
> I don't think mamba (or conda?) will work on Kaggle, but you can do pip installs

I've switched to `pip` for the `matplotx` install, but I can also try to add it to the kaggle environment by default.

As for the PUDL dependency, everything I'm importing here is just plotting functions that we could also just copy and paste into the notebook. It might not be ideal as it's a lot to read through, but also it would definitely be the easiest solution at this point.

> Is there a reason not to rely on the shorter and more readable S3 URLs rather than the https version?

Switched to s3 in the code + markdown blocks.

>  Referenced a different plot?

Ah yeah, I used the random state to try to make this stable, but over time I imagine that's not going to be good enough, so I'll just try to rework this to be more explicit.

> Our PG&E imputation is... wrong?

I definitely agree our version doesn't look great here. I think the start and end of the flagged section is at the bottom of the curve, and it seems to just impute values with a similar magnitude. Unfortunately, I kind of think that this is just a scenario where our imputation struggles, and I don't really have any ideas for "quick fixes".

As for other similar instances, I do think there's a pretty clear pattern you can see for PG&E in the correlation plot that I don't see anywhere else. That being said, I can do a little bit more exploration and see if I can identify any similar sections.

zschira commented:
@zaneselvans I've copied the plot functions into the notebook, so it no longer needs to import pudl.

As for the weird PG&E case, I've done more exploration to try to and identify any other similar cases. I've done this mostly by using the `plot_correlation` function and limiting to various BA's/subregions/timeframes to find sections where the correlation looked less close. I then actually plotted the imputation for those timeframes, but everything I've looked at seems consistent with other examples we've reviewed. Sometimes there's differences between the versions, but I've never seen anything where our version is so obviously off.

I've added a note in the notebook that this looks concerning and that we haven't been able to identify any other similar sections. At this point, I'm not sure if we want to invest more time in figuring this out or just accept that it's a limitation.

zaneselvans commented:
Running the notebook locally, I'm finding some parts are taking quite a while -- in particular the cell that pulls down the Zenodo archive w/ Tyler & Alicia's imputation in it has been running for 17 minutes already. Not sure if my internet is just particularly slow today or what, but if we expect any cell to take more than a minute to execute we probably want to mention that in the the notebook so folks don't wonder if something is broken.

zaneselvans commented:
Notes from check-in / review call:

* Make it clearer that the only difference between our data product and Tyler/Alicia's is the imputation method.  We use the same flagging (developed by Tyler), and the same EIA-930 soruce data.
* Explain why we didn't just use their imputation method (ours is much faster, and we want to reprocess the data regularly as it comes out)
* We should be comparing both subregion AND BA level data against Tyler / Alicia's outputs.  As a result of our inner merge early on that uses both BA codes and subregion codes, we toss all the BA-only data.
* We should have a little up-front "In general, this is what the data is." and "In general this is imputation was good!"
* We can't show the simulation outputs because we haven't worked them into the nightly buid outputs, but we can talk about what we found.
* Do better hand selection of the examples of imputations that we're going to use as demonstration -- don't use the tiny BAs or subregions.
* Also need to make sure that you try and find anomalies in *all the BA data** as well -- not just the subregion data.
